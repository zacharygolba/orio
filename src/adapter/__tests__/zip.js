// @flow

import Zip from '../zip'
import * as iter from '../../'

let subj

beforeEach(() => {
  const producerA = iter.repeat('test').producer
  const producerB = iter.from([1, 2, 3]).producer
  subj = new Zip(producerA, producerB)
})

test('#@@iterator()', () => {
  for (const item of subj) {
    expect(item).toMatchSnapshot()
  }
})

describe('#next()', () => {
  test('with an unbound lhs and a bound rhs', () => {
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
  })

  test('with a bound lhs and an unbound rhs', () => {
    const producerA = iter.from([1, 2, 3]).producer
    const producerB = iter.repeat('test').producer
    subj = new Zip(producerA, producerB)

    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
  })

  test('with a self reference to bound lhs', () => {
    const { producer } = iter.from([1, 2, 3, 4])
    subj = new Zip(producer, producer)

    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
    expect(subj.next()).toMatchSnapshot()
  })
})

describe('#sizeHint()', () => {
  test('with an unbound lhs and a bound rhs', () => {
    expect(subj.sizeHint()).toEqual(3)
  })

  test('with a bound lhs and an unbound rhs', () => {
    const producerA = iter.from([1, 2, 3]).producer
    const producerB = iter.repeat('test').producer
    subj = new Zip(producerA, producerB)

    expect(subj.sizeHint()).toEqual(3)
  })

  test('with two unbound producers', () => {
    const producerA = iter.repeat('test').producer
    const producerB = iter.repeat('test').producer
    subj = new Zip(producerA, producerB)

    expect(subj.sizeHint()).toEqual(Infinity)
  })

  test('with a self reference to bound lhs', () => {
    {
      const { producer } = iter.from([1, 2, 3])
      subj = new Zip(producer, producer)
    }

    expect(subj.sizeHint()).toBe(1)

    {
      const { producer } = iter.from([1, 2, 3, 4])
      subj = new Zip(producer, producer)
    }

    expect(subj.sizeHint()).toBe(2)
  })

  test('with a self reference to an unbound lhs', () => {
    const { producer } = iter.repeat('test')
    subj = new Zip(producer, producer)

    expect(subj.sizeHint()).toBe(Infinity)
  })
})
