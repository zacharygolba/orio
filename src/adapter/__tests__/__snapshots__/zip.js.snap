// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#@@iterator() 1`] = `
Array [
  "test",
  1,
]
`;

exports[`#@@iterator() 2`] = `
Array [
  "test",
  2,
]
`;

exports[`#@@iterator() 3`] = `
Array [
  "test",
  3,
]
`;

exports[`#next() with a bound lhs and an unbound rhs 1`] = `
Object {
  "done": false,
  "value": Array [
    1,
    "test",
  ],
}
`;

exports[`#next() with a bound lhs and an unbound rhs 2`] = `
Object {
  "done": false,
  "value": Array [
    2,
    "test",
  ],
}
`;

exports[`#next() with a bound lhs and an unbound rhs 3`] = `
Object {
  "done": false,
  "value": Array [
    3,
    "test",
  ],
}
`;

exports[`#next() with a bound lhs and an unbound rhs 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`#next() with a self reference to bound lhs 1`] = `
Object {
  "done": false,
  "value": Array [
    1,
    2,
  ],
}
`;

exports[`#next() with a self reference to bound lhs 2`] = `
Object {
  "done": false,
  "value": Array [
    3,
    4,
  ],
}
`;

exports[`#next() with a self reference to bound lhs 3`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`#next() with a self reference to bound lhs 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`#next() with an unbound lhs and a bound rhs 1`] = `
Object {
  "done": false,
  "value": Array [
    "test",
    1,
  ],
}
`;

exports[`#next() with an unbound lhs and a bound rhs 2`] = `
Object {
  "done": false,
  "value": Array [
    "test",
    2,
  ],
}
`;

exports[`#next() with an unbound lhs and a bound rhs 3`] = `
Object {
  "done": false,
  "value": Array [
    "test",
    3,
  ],
}
`;

exports[`#next() with an unbound lhs and a bound rhs 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;
